cmd/zoekt-sourcegraph-indexserver/owner.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"time"
+)
+
+type ownerChangeError struct {
+	Path           string
+	Owner, Current string
+}
+
+func (e *ownerChangeError) Error() string {
+	return fmt.Sprintf("detected a change of ownership of %s. In multiple replica setups this can lead to un-needed rebalancing or bugs if there are multiple writers: owner=%q current=%q", e.Path, e.Owner, e.Current)
+}
+
+// ownerChecker can write and check the owner file for a index directory. It
+// is used for detecting when multiple zoekts are writing to a directory, or
+// when the ownership is changing too often.
+//
+// The motivation for this is a person can misconfigure zoekt such that
+// multiple indexservers write to the same directory. This will lead to index
+// thrashing and hard to debug errors. Alternatively if the stable identity
+// (hostname) changes, this can lead to Sourcegraph's repo <-> owner hash
+// changing which means unnecessary rebalancing.
+type ownerChecker struct {
+	Path     string
+	Hostname string
+}
+
+// Run will regularly init then regularly check if we are owner. If an error
+// is detected it will exit the current process with exit code 1. This method
+// blocks.
+func (o *ownerChecker) Run() {
+	if err := o.Init(); err != nil {
+		log.Fatal(err)
+	}
+	for {
+		time.Sleep(5 * time.Second)
+		if err := o.Check(); err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func (o *ownerChecker) Init() error {
+	var ownerErr *ownerChangeError
+	if err := o.Check(); errors.Is(err, fs.ErrNotExist) {
+		// do nothing, first run so we just write out the file
+	} else if errors.As(err, &ownerErr) {
+		debug.Printf("WARN: detected a change in ownership at startup. You can ignore this if you only have one zoekt replica: %s", err)
+	} else if err != nil {
+		return err
+	}
+
+	content := []byte(fmt.Sprintf(`DO NOT EDIT! generated by zoekt-sourcegraph-indexserver.
+This file records the identity of the owner of this zoekt index directory.
+If it changes zoekt-sourcegraph-indexserver will exit with a non-zero exit code.
+This is to prevent multiple owners/writers.
+
+hostname=%s
```

Comment:
nit: make it the first line?

Reply:
out of interest, why? I chose last since when you cat a file this is the easiest thing to see. But given the file is small maybe the first line makes sense. WDYT? cc @stefanhengl wdyt?